logger/logger.go

```diff
@@ -1,120 +1,79 @@
+	e.data = append(e.data, l.Spacer...)
+	if l.Timestamp != "" {
+		e.data = time.Now().AppendFormat(e.data, l.Timestamp)
+	}
+	e.data = append(e.data, l.Spacer...)
+	e.data = append(e.data, []byte(data)...)
+	if data == "" || data[len(data)-1] != '\n' {
+		e.data = append(e.data, '\n')
+	}
+	e.output.Write(e.data)
+	e.data = (e.data)[:0]
+	eventPool.Put(e)
 
-	warnLogger = log.New(os.Stdout,
-		"[WARN]:  ",
-		log.Ldate|log.Ltime)
+	return nil
+}
 
-	errorLogger = log.New(os.Stdout,
-		"[ERROR]: ",
-		log.Ldate|log.Ltime)
+// CloseLogger is called on shutdown of application
+func CloseLogger() error {
+	err := GlobalLogFile.Close()
+	if err != nil {
+		return err
+	}
+	return nil
 
-	fatalLogger = log.New(os.Stdout,
-		"[FATAL]: ",
-		log.Ldate|log.Ltime)
 }
 
-// colorOutput() sets the prefix of each log type to matching colour
-// TODO: add windows support
-func colourOutput() {
-	if runtime.GOOS != "windows" || Logger.ColourOutputOverride {
-		debugLogger.SetPrefix("\033[34m[DEBUG]\033[0m: ")
-		infoLogger.SetPrefix("\033[32m[INFO]\033[0m: ")
-		warnLogger.SetPrefix("\033[33m[WARN]\033[0m: ")
-		errorLogger.SetPrefix("\033[31m[ERROR]\033[0m: ")
-		fatalLogger.SetPrefix("\033[31m[FATAL]\033[0m: ")
+func validSubLogger(s string) (bool, *subLogger) {
+	if v, found := subLoggers[s]; found {
+		return true, v
 	}
+	return false, nil
 }
 
-// clearAllLoggers() sets all logger flags to 0 and outputs to Discard
-func clearAllLoggers() {
-	debugLogger.SetFlags(0)
-	infoLogger.SetFlags(0)
-	warnLogger.SetFlags(0)
-	errorLogger.SetFlags(0)
-	fatalLogger.SetFlags(0)
+func Level(s string) (*Levels, error) {
+	found, logger := validSubLogger(s)
+	if !found {
+		return nil, fmt.Errorf("logger %v not found", logger)
+	}
 
-	debugLogger.SetOutput(ioutil.Discard)
-	infoLogger.SetOutput(ioutil.Discard)
-	warnLogger.SetOutput(ioutil.Discard)
-	errorLogger.SetOutput(ioutil.Discard)
-	fatalLogger.SetOutput(ioutil.Discard)
+	return &logger.Levels, nil
 }
 
-// setupOutputs() sets up the io.writer to use for logging
-// TODO: Fix up rotating at the moment its a quick job
-func setupOutputs() (err error) {
-	if len(Logger.File) > 0 {
-		logFile := filepath.Join(LogPath, Logger.File)
-		if Logger.Rotate {
-			if _, err = os.Stat(logFile); !os.IsNotExist(err) {
-				currentTime := time.Now()
-				newName := currentTime.Format("2006-01-02 15-04-05")
-				newFile := newName + " " + Logger.File
-				err = os.Rename(logFile, filepath.Join(LogPath, newFile))
-				if err != nil {
-					err = fmt.Errorf("failed to rename old log file %s", err)
-					return
-				}
-			}
-		}
-		logFileHandle, err = os.OpenFile(logFile, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
-		if err != nil {
-			return
-		}
-		logOutput = io.MultiWriter(os.Stdout, logFileHandle)
-	} else {
-		logOutput = os.Stdout
+func SetLevel(s, level string) (*Levels, error) {
+	found, logger := validSubLogger(s)
+	if !found {
+		return nil, fmt.Errorf("logger %v not found", logger)
```

Comment:
Since the pointer is passed back from validSubLoggers there is almost no performance benefit for or against having it return a bool and comes down to code complexity/readability

Is it easier to return a bool, pointer and check if the bool is true/false or return a pointer and check if its nil?

BenchmarkBothValidSubLoggers/NoBool-8         	  500000	      3389 ns/op	    1243 B/op	      16 allocs/op
BenchmarkBothValidSubLoggers/Bool-8           	  500000	      3386 ns/op	    1304 B/op	      16 allocs/op

Reply:
Up to you.